RSE Manager: retry stat multiple times before failing Fix #1833

berghaus opened the pull request:
RSE Manager: retry stat multiple times before failing Fix #1833

Commit:
RSE Manager: retry stat multiple times before failing

When requesting file stats of a storage element immediately on the
competion of the upload sometimes this fails. The storage element may
not yet have completed internal operations causing a race condition. To
avoid this retry again with an increasing pause inbetween. After a
minute assume that the storage element should have completed it's
internal operations.

mlassnig commented on lib/rucio/rse/rsemanager.py:
```diff
@@ -703,6 +703,26 @@ def find_matching_scheme(rse_settings_dest, rse_settings_src, operation_src, ope
+    :param protocol     The protocol to use to reach this file
+    :param pfn          Physical file name of the target for the protocol stat
+    """
+    from time import sleep
```
imports always go to the top of the module

bari12 replied:
Pinging @berghaus 😄

bari12 commented on lib/rucio/rse/rsemanager.py:
```diff
@@ -703,6 +703,26 @@ def find_matching_scheme(rse_settings_dest, rse_settings_src, operation_src, ope
+    :param pfn          Physical file name of the target for the protocol stat
+    """
+    from time import sleep
+    for attempt in range(6):
```
In the baseclient there is a variable called `REQUEST_RETRIES`, I wonder if we should use this here as well.

berghaus replied:
Looking through this it is specified in the Rucio configuration and used in the loop in line [298](https://github.com/rucio/rucio/blob/6d17a9dac8fb2844c10342dd74d39e9ff0f6161b/lib/rucio/client/baseclient.py#L298) of the baseclient. It sets a number of times to try opening a connection.

Using it here would overload the meaning: here the connection can be opened but it may take the storage element some time to recognise the data to be there. We could use this variable here, but should we be worried about introducing some interplay between what it does to connection retries and what it does in the RSE manager?

mlassnig committed:
dependencies: remove pycrypto; Fix #1475

mlassnig committed:
messed up rebase

bari12 committed:
Core: Increased size of extended_attributes column in rse_protocols; Fix #1543

bari12 committed:
Release: Added alembic script for missing comments column in subscriptions_history; Fix #2238

bari12 committed:
Core: Fixing multiple alembic heads

berghaus commented:
The commit has been rebased on the current next branch. Let me know what you think about my previous question.

bari12 committed:
Merge pull request #2222 from cserf/patch-2221-Minos_is_not_able_to_handle_PFNs_with_different_schemes

Minos is not able to handle PFNs with different schemes : Closes #2221

bari12 committed:
Merge pull request #2259 from TWAtGH/patch-2141-generic_metadata_fix

Core & Internals: generic metadata fix and small client cli cleanup; Fix #2141

Commit:
RSE Manager: make the number of stat retries configurable Fix #1833

Configure the number of times the RSE manager will attempt to stat a
file before giving up with protocol_stat_retries in the client section.

berghaus commented:
Added a new configuration variable `protocol_stat_retries` set to 6 by default and defaulting to 6 if not picked up from the configuration. The forced push was necessary because of another rebase.

berghaus commented:
Replaced this feature by a patch since we agreed that by now this has become too urgent to wait until June.
